### PR TITLE
fix: prevent duplicate heading IDs across prose blocks breaking TOC links

### LINKED ISSUE #1952: [DETAIL] Table of contents links break due to duplicate auto-generated heading IDs across prose blocks
> [!NOTE]
> From https://app.detail.dev/oss/opengovsg/isomer/bug_9b56534b-b9d6-4106-b1f9-ec53b01d301c
 
# Summary
- **Context**: The `getDigestFromText` function generates MD5 hash IDs used for table of contents anchor links and block identifiers.
- **Bug**: IDs for prose level-2 headings are generated without the outer block index, causing duplicate IDs when multiple prose blocks contain identical headings at the same position.
- **Actual vs. expected**: Actual: identical level-2 headings in different prose blocks can receive the same `id`, producing duplicate HTML IDs and TOC links that both point to the first heading. Expected: each heading should receive a unique `id` so each TOC entry navigates to its corresponding heading.
- **Impact**: Invalid HTML (duplicate IDs), broken table of contents anchor links that only navigate to the first matching element, and accessibility violations.

# Code with Bug
```typescript
// packages/components/src/utils/getTransformedPageContent.ts
const anchorId = getDigestFromText(
  `${JSON.stringify(component)}_${componentIndex}`,
) // <-- BUG 🔴 missing outer prose-block index, so headings in different blocks can collide
```

# Explanation
`getTransformedPageContent` auto-generates IDs only for level-2 headings that don’t already have an `id`. The hash input includes only the serialized heading node and its position within a single prose block (`componentIndex`). If multiple prose blocks contain the same heading structure at the same inner index (e.g., both blocks start with an `h2` “Introduction”), the MD5 inputs are identical, so the IDs collide. `getTableOfContents` then builds `anchorLink: "#" + component.attrs.id`, so multiple TOC entries can end up linking to the same DOM element (the first matching ID).

## Codebase Inconsistency
Complex blocks already include the outer block index when generating IDs, avoiding collisions:
```typescript
// packages/components/src/utils/getTransformedPageContent.ts
return {
  ...block,
  id: getDigestFromText(`${JSON.stringify(block)}_${index}`),
}
```

# Failing Test
```typescript
// packages/components/src/utils/__tests__/getTransformedPageContent.test.ts
it("generates unique IDs for identical headings in different prose blocks", () => {
  const content: IsomerComponent[] = [
    {
      type: "prose",
      content: [
        {
          type: "heading",
          attrs: { level: 2 },
          content: [{ type: "text", text: "Introduction" }],
        },
      ],
    },
    {
      type: "prose",
      content: [
        {
          type: "heading",
          attrs: { level: 2 },
          content: [{ type: "text", text: "Introduction" }],
        },
      ],
    },
  ]
  const transformed = getTransformedPageContent(content)
  const heading1 = transformed[0]?.content?.[0]
  const heading2 = transformed[1]?.content?.[0]
  expect(heading1?.attrs?.id).toMatch(idPattern)
  expect(heading2?.attrs?.id).toMatch(idPattern)
  expect(heading1?.attrs?.id).not.toBe(heading2?.attrs?.id)
})
```

Test output:
```
FAIL  src/utils/__tests__/getTransformedPageContent.test.ts > getTransformedPageContent > generates unique IDs for identical headings in different prose blocks
AssertionError: expected '9c479136a83119bff5c9bdc9b6bb9424' not to be '9c479136a83119bff5c9bdc9b6bb9424' // Object.is equality
```

# Recommended Fix
```typescript
// packages/components/src/utils/getTransformedPageContent.ts
const anchorId = getDigestFromText(
  `${JSON.stringify(component)}_${index}_${componentIndex}`,
)
```

# History
This bug was introduced in commit e0fd6bb3.

### PR DESCRIPTION
**Pull request by jovnc**
## Problem

Closes #1952

Level-2 prose headings get auto-generated ID whose MD5 hash input can be the same, if different prose blocks have identical headings. This causes duplicate HTML ID and ToC links to jump only to the first matching heading.

## Solution

* Include the outer block `index` in the hash input, similar to how `infocards`, `infocols`, etc. generate their IDs.
* For backwards compatibility, only duplicated prose block with identical headings (2nd block onwards) will have `index` appended to hash input to ensure old links still work

**Breaking Changes**

- [x] Yes - this PR contains breaking changes
  - Auto-generated heading IDs change, so old external/bookmarked anchor links (`#<hash>`) may break. On-page TOC links are unaffected as they regenerate alongside the IDs on each build (mitigated in solution above).
- [ ] No - this PR is backwards compatible

**Features**:

- NIL

**Improvements**:

- NIL

**Bug Fixes**:

- Fix duplicate auto-generated heading IDs across prose blocks that broke TOC anchor links and produced invalid HTML.

## Before & After Screenshots

**BEFORE**:

NIL

**AFTER**:

NIL

## Tests

**Manual Verification Steps**:

- [ ] Create a page with two prose blocks, each starting with an identical level-2 heading (e.g. "Introduction").
- [ ] Confirm both headings render with distinct `id` attributes.
- [ ] Click each TOC entry and confirm it scrolls to its own heading (not just the first).
- [ ] Run the unit test `pnpm test:unit -- src/utils/__tests__/getTransformedPageContent.test.ts`

**New scripts**:

NIL

**New dependencies**:

NIL

**New dev dependencies**:

NIL

### CODE REVIEW

File: packages/components/src/utils/getTransformedPageContent.ts [hunk 1]
```diff
@@ -19,7 +19,7 @@ export const getTransformedPageContent = (
           ) {
             // generate a unique hash to auto-generate anchor links
             const anchorId = getDigestFromText(
-              `${JSON.stringify(component)}_${componentIndex}`,
+              `${JSON.stringify(component)}_${index}_${componentIndex}`,
```
[the remark this answers is not in the document]
Comment (jovnc) [reply packages/components/src/utils/getTransformedPageContent.ts]: Should follow existing patterns, rather than create new approach here.

[the remark this answers is not in the document]
Comment (jovnc) [reply packages/components/src/utils/getTransformedPageContent.ts]: This is a valid concern, but handling it via collision detection makes the solution noticeably more convoluted and feels more like a patch than a fix.

@adriangohjw What do you think about this?

[the remark this answers is not in the document]
Comment (adriangohjw) [reply packages/components/src/utils/getTransformedPageContent.ts]: the review has a point and ideally we maintain backward compatibility now, mainly because we are unsure of how the vast end users are using this for now. can we look into that?

**Review by adriangohjw — APPROVED** [review]
thanks! i think this implementation is good and simple

**Review by adriangohjw — COMMENTED** [review]
## Summary

- Fixes #1952 by deduplicating auto-generated heading anchor IDs across prose blocks. Previously, two prose blocks each starting with `## Introduction` produced the same MD5 (seed was `JSON.stringify(component)_componentIndex`, both 0), so TOC anchors all jumped to the first heading and HTML had duplicate IDs.
- Codeflow: `getTransformedPageContent.ts:10-34` walks the content array; for each level-2 prose heading without an explicit `id`, it generates the legacy seed (`packages/components/src/utils/getTransformedPageContent.ts:23`) and, if the resulting hash is already in `usedHeadingIds`, re-hashes with the outer block `index` appended (`getTransformedPageContent.ts:25-26`). `getTableOfContents.ts:34` then consumes those IDs unchanged.
- Overall: small, surgical change with reasonable backwards-compat behaviour (first occurrence keeps its legacy hash, only 2nd+ duplicates shift). Implementation is simple and the regression test pins the legacy-seed contract. A couple of edge cases worth considering below.

## Findings

**issue (correctness, non-blocking):** `packages/components/src/utils/getTransformedPageContent.ts:25-28` — the rehashed value `getDigestFromText(`${seed}_${index}`)` is not itself checked against `usedHeadingIds` before being added. In practice MD5 collisions are vanishingly unlikely, but more importantly this also doesn't protect against the case where the *rehashed* value of one heading happens to match the *legacy* hash of a future heading. The simple form is a `while (usedHeadingIds.has(anchorId)) { anchorId = getDigestFromText(`${seed}_${index}_${attempt++}`) }` loop. Given MD5's space this is theoretical, but the code currently presents itself as a dedup guarantee when it's really a 1-pass best-effort.

Confidence: medium

**question:** `getTransformedPageContent.ts:25` — using the outer block `index` means that if an editor later reorders prose blocks (without changing any heading text), the 2nd-and-onwards duplicate heading IDs will shift to whatever block index they now occupy. The first occurrence is stable, but the 2nd+ are not. The PR body mentions IDs change once at rollout — is the ongoing reorder sensitivity also acceptable, or worth recording somewhere? An alternative seed (e.g. a running duplicate counter independent of block position) would be order-stable.

Confidence: high

**suggestion (if-minor):** `packages/components/src/utils/__tests__/getTransformedPageContent.test.ts:92-119` — the new test only covers two identical headings at `componentIndex === 0`. Worth adding one or both of:
  - 3+ identical prose blocks (verifies the loop handles N > 2, currently relies on each iteration producing a unique `index`).
  - Duplicate heading at a non-zero `componentIndex` (e.g. heading preceded by a paragraph in both prose blocks) to assert the `componentIndex` part of the seed still threads through.

These would harden the regression guard against future refactors of the seed format.

Confidence: high

**nitpick:** `getTransformedPageContent.ts:22` — the inline comment is ~150 chars on one line and reads awkwardly. Wrapping to two lines or shortening to "only re-hash with block index when this heading's legacy hash collides — preserves legacy IDs for non-duplicated headings" would read better.

Confidence: low

**note:** the `usedHeadingIds` set only tracks auto-generated hashes, so a heading whose author-supplied `id` happens to equal a generated hash will still produce duplicate DOM IDs. Pre-existing behaviour, not introduced here — flagging only in case it's worth a follow-up.

Confidence: medium

**praise:** the backwards-compat decision to keep the first occurrence on the legacy seed (`getTransformedPageContent.ts:23`) is the right call — minimises broken bookmarks for the common single-heading case.

Confidence: high

**Comment by adriangohjw** [comment]
@jovnc can ignore the issue/questions for the above automated review, but i think the "suggestion" and "nitpick" are worth working on. Can you work on them according? i will merge them once done, thanks!

**Comment by jovnc** [comment]
@adriangohjw thanks for the review! address the suggestion and nitpick accordingly
